src/utils/size-observer.js

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {pushIfNotExist, removeItem} from './array';
+import {rethrowAsync} from '../log';
+
+/** @typedef {function(../layout-rect.LayoutSizeDef)} */
+let ContentSizeObserverCallbackDef;
+
+/** @const {!WeakMap<!Window, !ResizeObserver>} */
+const observers = new WeakMap();
+
+/** @const {!WeakMap<!Element, !ContentSizeObserverCallbackDef>} */
```

Comment:
Should be `WeakMap<!Element, !Array<!ContentSizeObserverCallbackDef>>` right?  We want the value to be a list of callbacks.

The name I've heard before for this structure is a [Multimap](https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Multimap.html).

Nit: can we improve these two variable names. `targetObserverListMap` doesn't indicate it holds callbacks and `targetSizeListMap` isn't a "list map" (it holds only a single size per element). Maybe `callbackMultimap` / `elementSizeMap`.

Reply:
Fixed and renamed to multimap. Surprised the compiler didn't error here.